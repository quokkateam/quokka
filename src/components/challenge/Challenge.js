import React, { Component } from 'react';

import Ajax from '../../utils/Ajax';
import BannerSection from './BannerSection';
import ChallengeSection  from './ChallengeSection';
import CheckInSection from './CheckInSection';
import ExtrasSection from './ExtrasSection';
import GettingStartedSection  from './GettingStartedSection';
import InviteFriendsSection from './InviteFriendsSection';
import OverviewSection from './OverviewSection';
import PrizesSection from './PrizesSection';

class Challenge extends Component {

  constructor(props) {
    super(props);
    
    this.weekNum = this.props.match.params.weekNum;

    this.setChallengeSectionRef = this.setChallengeSectionRef.bind(this);
    this.setPrizesSectionRef = this.setPrizesSectionRef.bind(this);
    this.setSuggestionsSectionRef = this.setSuggestionsSectionRef.bind(this);
    this.setExtrasSectionRef = this.setExtrasSectionRef.bind(this);
    this.onChallengeSectionUpdate = this.onChallengeSectionUpdate.bind(this);
    this.handleChallengeFetched = this.handleChallengeFetched.bind(this);

    this.state = {
      challengeId: null,
      habit: null,
      adjHabits: null,
      overview: null,
      challenge: null,
      prizes: null,
      sponsors: null,
      suggestions: null,
      links: null,
      extraInfo: null
    };
  }

  setChallengeSectionRef(ref) {
    this.challengeSection = ref;
  }

  setPrizesSectionRef(ref) {
    this.prizesSection = ref;
  }

  setSuggestionsSectionRef(ref) {
    this.suggestionsSection = ref;
  }

  setExtrasSectionRef(ref) {
    this.extrasSection = ref;
  }

  componentDidUpdate() {
    if (this.state.habit) {
      this.challengeSection.setState({
        challenge: this.state.challenge,
        challengeId: this.state.challengeId
      });

      this.prizesSection.setState({
        challengeId: this.state.challengeId,
        prizes: this.state.prizes,
        sponsors: this.state.sponsors,
        points: this.state.challenge.points || 0
      });

      this.suggestionsSection.setState({
        suggestions: this.state.suggestions,
        challengeId: this.state.challengeId
      });

      this.extrasSection.setState({
        links: this.state.links,
        extraInfo: this.state.extraInfo
      });
    }

    return true;
  }

  componentDidMount() {
    Ajax.get('/api/challenge/' + this.weekNum)
      .then((resp) => {
        this.handleChallengeFetched(resp);
      });
  }

  handleChallengeFetched(resp) {
    if (resp.status === 200 || resp.status === 400) {
      resp.json().then((data) => {
        if (resp.status === 200) {
          this.setState({
            challengeId: Number(data.id),
            habit: data.habit,
            adjHabits: data.adjHabits,
            overview: data.overview,
            challenge: data.challenge,
            prizes: data.prizes,
            sponsors: data.sponsors,
            suggestions: data.suggestions,
            links: data.links,
            extraInfo: data.extraInfo
          });
        } else {
          window.location = '/challenges';
        }
      });
    } else {
      console.warn('Unexpected response code while fetching challenge: ', resp.status);
    }
  }

  onChallengeSectionUpdate(challenge) {
    this.prizesSection.setState({ points: challenge.points || 0 });
  }

  render() {
    return (
      <div>
        <div id="challenge">
          <BannerSection habit={this.state.habit} weekNum={this.weekNum} adjHabits={this.state.adjHabits} />
          <OverviewSection overview={this.state.overview} />
          <ChallengeSection challenge={this.state.challenge} onUpdate={this.onChallengeSectionUpdate} ref={this.setChallengeSectionRef} />
          <PrizesSection prizes={this.state.prizes} sponsors={this.state.sponsors} points={(this.state.challenge || {}).points} ref={this.setPrizesSectionRef} />
          <GettingStartedSection suggestions={this.state.suggestions} ref={this.setSuggestionsSectionRef} />
          <CheckInSection weekNum={this.weekNum} endDate={((this.state.habit || {}).dates || {}).end} />
          <ExtrasSection links={this.state.links} extraInfo={this.state.extraInfo} ref={this.setExtrasSectionRef} />
        </div>
        <InviteFriendsSection />
      </div>
    );
  }
}

export default Challenge;
