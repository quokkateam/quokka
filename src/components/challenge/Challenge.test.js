import Challenge from './Challenge';
import { MemoryRouter } from 'react-router-dom';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<Challenge />', () => {
  it('renders correctly', () => {
    const challenge = renderer.create(
      <MemoryRouter>
        <Challenge match={{ params: { weekNum: 1} }} />
      </MemoryRouter>
    ).toJSON();
    expect(challenge).toMatchSnapshot();
  });
});