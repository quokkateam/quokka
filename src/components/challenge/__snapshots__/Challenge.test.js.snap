// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Challenge /> renders correctly 1`] = `
<div>
  <div
    id="challenge"
  >
    <div
      id="challengeBanner"
    >
      <div
        className="dimmer"
      />
      <div
        className="adj-week-nav prev-week"
      >
        <a
          href="/challenge"
          onClick={[Function]}
        >
          <div
            className="quokka-nav-button nav-left banner-nav-button"
          />
          <span
            className="adj-challenge-title"
          >
            Sleep
          </span>
        </a>
      </div>
      <div
        className="adj-week-nav next-week"
      >
        <a
          href="/challenge"
          onClick={[Function]}
        >
          <div
            className="quokka-nav-button nav-right banner-nav-button"
          />
          <span
            className="adj-challenge-title"
          >
            Journaling
          </span>
        </a>
      </div>
      <div
        className="main-content"
      >
        <img
          alt=""
          className="habit-icon"
          src="https://s3-us-west-1.amazonaws.com/quokkadev/images/running-green.png"
        />
        <div
          className="habit-name"
        >
          Exercise
        </div>
        <div
          className="week-date-info"
        >
          <span>
            Week 
            4
          </span>
          ·
          <span>
            Oct 15
          </span>
          —
          <span>
            Oct 22
          </span>
        </div>
      </div>
    </div>
    <div
      className="container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="challenge-section-title"
        >
          Habit Overview
        </div>
        <div
          className="challenge-section-desc"
        >
          <div
            className="overview-subsection"
          >
            <div
              className="challenge-ov-subtitle"
              onClick={[Function]}
            >
              What’s the science say?
            </div>
            <div
              className="challenge-ov-content collapse"
            >
              <p>
                Regular physical activity is important for increasing energy and focus,
                strengthening bones and muscles, and increasing longevity
                
                <a
                  href="https://www.cdc.gov/physicalactivity/basics/pa-health/index.htm#StrengthenBonesMuscle"
                  title={undefined}
                >
                  [1]
                </a>
                .
                It'll also help you manage stress and anxiety and reduce your risk of
                cardiovascular disease and obesity. Even our perceptions and mindsets on
                how physically active we are can predict health and longevity in the
                domains of stress, diet, and obesity
                
                <a
                  href="http://news.stanford.edu/2017/07/20/self-perceptions-linked-shorter-lifespans/"
                  title={undefined}
                >
                  [2]
                </a>
                .
              </p>
            </div>
          </div>
          <div
            className="overview-subsection"
          >
            <div
              className="challenge-ov-subtitle"
              onClick={[Function]}
            >
              How much exercise should I get?
            </div>
            <div
              className="challenge-ov-content collapse"
            >
              <p>
                The Physical Activity Guidelines for Americans recommends 2.5 hours per
                week of moderate-intensity aerobic activity (or 75 minutes of more
                vigorous exercise) for good general health. Depending on how intense the
                aerobic activity, you could take a brisk 30-min walk 5x a week or play a
                high-intensity sport for a shorter period
                
                <a
                  href="http://www.health.gov/paguidelines/"
                  title={undefined}
                >
                  [3]
                </a>
                . Adding in strength
                training twice a week can help with weight loss and increase your
                resting metabolic rate on top of all this!
              </p>
            </div>
          </div>
          <div
            className="overview-subsection"
          >
            <div
              className="challenge-ov-subtitle"
              onClick={[Function]}
            >
              What are some easy things to do?
            </div>
            <div
              className="challenge-ov-content collapse"
            >
              <p>
                There are easy ways to incorporate physical activity in your day-to-day.
                Sitting for prolonged periods throughout the day has been linked to
                heart disease and diabetes, and stepping away from the chair or even
                just standing for a few minutes every hour can help reduce these
                negative effects
                
                <a
                  href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3404815/"
                  title={undefined}
                >
                  [4]
                </a>
                ,
                
                <a
                  href="https://www.ncbi.nlm.nih.gov/pubmed/25931456"
                  title={undefined}
                >
                  [5]
                </a>
                . Choosing to walk
                or bike, taking the stairs, and cleaning your room are easy ways to
                incorporate more physical activity in your day.
              </p>
            </div>
          </div>
          <div
            className="overview-subsection"
          >
            <div
              className="challenge-ov-subtitle"
              onClick={[Function]}
            >
              Should I focus on strength or cardio?
            </div>
            <div
              className="challenge-ov-content collapse"
            >
              <p>
                It’s important to do both! Strength training is important for building
                muscles, preventing injury, and improving your metabolism. Cardio
                (aerobic exercise that elevates your heart rate) improves heart health,
                helps with weight loss, and increases your metabolism, too
                
                <a
                  href="https://health.clevelandclinic.org/2016/02/head-toe-benefits-cardio-workout-infographic/"
                  title={undefined}
                >
                  [6]
                </a>
                .
              </p>
            </div>
          </div>
          <div
            className="overview-subsection"
          >
            <div
              className="challenge-ov-subtitle"
              onClick={[Function]}
            >
              If you’re getting bored, switch it up!
            </div>
            <div
              className="challenge-ov-content collapse"
            >
              <p>
                Try mixing up your workout—varying your exercise routines will help you
                avoid plateaus in workout performance and also improve adherence.
                Figuring out what activities get you most excited and motivated will be
                key to getting consistent physical activity.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="container-fluid challenge-card"
    >
      <div
        className="row"
      >
        <div
          className="challenge-card-header"
        >
          <div
            className="challenge-section-title"
          >
            Challenge of the Week
          </div>
          <div
            className="points"
          >
            <img
              alt=""
              src="https://s3-us-west-1.amazonaws.com/quokkadev/images/leaf.png"
            />
            <span
              className="point-count"
            >
              60
            </span>
          </div>
        </div>
        <div
          className="challenge-card-body"
        >
          <div>
            <ol
              className={undefined}
              data-sourcepos={undefined}
            >
              <li>
                Try 3 different types of exercise this week
              </li>
              <li>
                Walk or bike instead of driving, and take stairs over elevators
              </li>
              <li>
                Sit Less, Move More
              </li>
            </ol>
          </div>
        </div>
      </div>
    </div>
    <div
      className="container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="challenge-section-title"
        >
          Prizes
        </div>
        <div
          className="challenge-section-desc"
        >
          Participants of this week’s challenge will earn 
          <span
            className="featured"
          >
            60 Quokka points
          </span>
           and be eligible for the following prizes:
        </div>
        <ul
          className="prizes"
        >
          <li>
            <a
              href="https://thenorthface.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                className="prize"
              >
                <img
                  alt=""
                  className="sponsor-icon"
                  src="https://s3-us-west-1.amazonaws.com/quokkadev/images/thenorthface.png"
                />
                <div
                  className="prize-text"
                >
                  Free pair of shoes
                </div>
                <div
                  className="sponsor-name"
                >
                  The North Face
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="https://lululemon.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                className="prize"
              >
                <img
                  alt=""
                  className="sponsor-icon"
                  src="https://s3-us-west-1.amazonaws.com/quokkadev/images/lululemon.png"
                />
                <div
                  className="prize-text"
                >
                  Yoga mat giveaway
                </div>
                <div
                  className="sponsor-name"
                >
                  Lululemon
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="http://recreation.stanford.edu/"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                className="prize"
              >
                <img
                  alt=""
                  className="sponsor-icon"
                  src="https://s3-us-west-1.amazonaws.com/quokkadev/images/stanford-rec.png"
                />
                <div
                  className="prize-text"
                >
                  Free personal training session
                </div>
                <div
                  className="sponsor-name"
                >
                  Stanford Recreation
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="http://www.gostanford.com/"
              rel="noopener noreferrer"
              target="_blank"
            >
              <div
                className="prize"
              >
                <img
                  alt=""
                  className="sponsor-icon"
                  src="https://s3-us-west-1.amazonaws.com/quokkadev/images/stanford-rec.png"
                />
                <div
                  className="prize-text"
                >
                  Stanford swag
                </div>
                <div
                  className="sponsor-name"
                >
                  Stanford Athletics
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="challenge-section-title"
        >
          Need help getting started?
        </div>
        <ul
          className="suggestions-list"
        >
          <li>
            <div>
              <p>
                Sign up for a 
                <a
                  href="http://recreation.stanford.edu/fitness/assessment/"
                  title={undefined}
                >
                  Fitness
                  Assessment
                </a>
                 with
                Stanford Recreation.
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Jump into new 
                <a
                  href="http://recreation.stanford.edu/fitness/group/"
                  title={undefined}
                >
                  Group Fitness
                  classes
                </a>
                 like TRX,
                F45 Cardio, Cardio Kickboxing, or Barre Sculpt.
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Do yoga in Memorial Church as part of the monthly 
                <a
                  href="http://events.stanford.edu/events/704/70439/"
                  title={undefined}
                >
                  Om Under the
                  Dome
                </a>
                .
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Get together a group of friends, 
                <a
                  href="http://recreation.stanford.edu/membership/equipment/"
                  title={undefined}
                >
                  borrow equipment from the
                  gym
                </a>
                , and try
                sports like squash, racquetball, and badminton.
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Take free belay lessons, rent shoes for $2 from the Outdoor Center,
                and check out our 
                <a
                  href="http://recreation.stanford.edu/climbing/"
                  title={undefined}
                >
                  gym climbing and bouldering
                  hours
                </a>
                .
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Check out other recreation spaces beyond ACSR and AOERC – head to
                the 
                <a
                  href="https://golfcourse.stanford.edu/proshop.htm"
                  title={undefined}
                >
                  Stanford Driving
                  Range
                </a>
                 (only pay for
                the golf balls), 
                <a
                  href="https://dish.stanford.edu/"
                  title={undefined}
                >
                  hike the Dish
                </a>
                , or
                choose a 
                <a
                  href="https://www.walkjogrun.net/running-routes/USA/CA/Stanford/"
                  title={undefined}
                >
                  new campus running
                  trail
                </a>
                
                created by your peers!
              </p>
            </div>
          </li>
          <li>
            <div>
              <p>
                Seek a little inspiration from our top Stanford athletic teams with
                free 
                <a
                  href="http://gostanford.com/redzone"
                  title={undefined}
                >
                  Red Zone
                </a>
                 student tickets.
              </p>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="challenge-section-title"
        >
          Check-In
        </div>
        <div
          className="challenge-section-desc"
        >
          Don't forget to submit your check-in this week by Sunday, 
          Oct 22
          ! Check-ins help you earn Quokka points, provide us with feedback, and make you eligible for this week's prizes! 
          <a
            href="/check-in"
          >
            Take me to my Week 
            4
             Check-In.
          </a>
        </div>
      </div>
    </div>
    <div
      className="container-fluid"
      id="extras"
    >
      <div
        className="row"
      >
        <div
          className="challenge-section-title"
        >
          Extra Info
        </div>
        <ul
          className="challenge-section-desc"
        >
          <li>
            [
            1
            ]: 
            <a
              href="https://www.cdc.gov/physicalactivity/basics/pa-health/index.htm#StrengthenBonesMuscle"
            >
              https://www.cdc.gov/physicalactivity/basics/pa-health/index.htm#StrengthenBonesMuscle
            </a>
          </li>
          <li>
            [
            2
            ]: 
            <a
              href="http://news.stanford.edu/2017/07/20/self-perceptions-linked-shorter-lifespans/"
            >
              http://news.stanford.edu/2017/07/20/self-perceptions-linked-shorter-lifespans/
            </a>
          </li>
          <li>
            [
            3
            ]: 
            <a
              href="http://www.health.gov/paguidelines/"
            >
              http://www.health.gov/paguidelines/
            </a>
          </li>
          <li>
            [
            4
            ]: 
            <a
              href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3404815/"
            >
              https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3404815/
            </a>
          </li>
          <li>
            [
            5
            ]: 
            <a
              href="https://www.ncbi.nlm.nih.gov/pubmed/25931456"
            >
              https://www.ncbi.nlm.nih.gov/pubmed/25931456
            </a>
          </li>
          <li>
            [
            6
            ]: 
            <a
              href="https://health.clevelandclinic.org/2016/02/head-toe-benefits-cardio-workout-infographic/"
            >
              https://health.clevelandclinic.org/2016/02/head-toe-benefits-cardio-workout-infographic/
            </a>
          </li>
        </ul>
        <div>
          <p>
            If circumstances prevent you from fulfilling our outlined challenges,
            
            <a
              href="https://nccih.nih.gov/health/stress/relaxation.htm"
              title={undefined}
            >
              mind-body exercises may be helpful alternatives to managing a variety of
              health conditions such as anxiety, insomnia, or other chronic pain
              symptoms
            </a>
            .
          </p>
          <p>
            The information on this site is for general purposes only and is not
            intended or implied to be a substitute for professional medical advice,
            diagnosis, or treatment. Seek advice from your health care provider for
            your personal health needs.
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container-fluid challenge-invite-friends"
  >
    <div
      className="leading-contact-text"
    >
      Have friends who still want to join the Challenge?
      <br />
      Send them an invite.
    </div>
    <div
      className="row challenge-invite-friends-form"
    >
      <input
        className="form-input contact-field"
        defaultValue=""
        name="email"
        onKeyUp={[Function]}
        placeholder="Email Address"
        type="text"
      />
      <button
        className="quokka-nav-button nav-right submit-desktop"
        onClick={[Function]}
      />
    </div>
    <div
      className="submit-mobile-container"
    >
      <div
        className="lg-spinner-btn submit-mobile"
      >
        <button
          onClick={[Function]}
        >
          Submit
        </button>
        <div
          className="horiz-spinner"
        >
          <div
            className="bounce1"
          />
          <div
            className="bounce2"
          />
          <div
            className="bounce3"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
